jekyll/_cci2/using-dynamic-configuration.adoc

```diff
@@ -198,20 +201,15 @@ workflows:
-** The `build-service-1` job uses the `maven` orb to compile/install the service1 code, but skips any tests
-** The `build-service-2` job uses the `maven` orb to compile/install the service2 code, but skips any tests
-** The `run-integration-tests` job uses the `maven` orb to run any integration tests
-* Define four workflows, three of which are conditionally executed:
-** The `service-1` workflow triggers the `build-service-1` job when the pipeline parameter value mapped to run-build-service-1-job is set to `true`
-** The `service-2` workflow triggers the `build-service-2` job when the pipeline parameter value mapped to run-build-service-2-job is set to `true`
-** The `run-integration-tests` workflow will run if the `run-build-service-1-job` or `run-build-service-2-job` pipeline parameters have been updated to `true` based on the results of the `path-filtering` orb
-** The `check-updated-files` workflow will always run any time this pipeline is triggered
-
-See the `path-filtering` link:https://circleci.com/developer/orbs/orb/circleci/path-filtering[orb documentation] for more
-information on available elements and required parameters.
+* **Line 4**: Invoke the Maven orb so we can use it's preconfigured jobs
+* **Lines 8-14**: Define our two boolean pipeline parameters, the same parameters we have defined in the setup phase: `run-build-service-1-job` and `run-build-service-2-job`
+* Define three separate workflows to be conditionally executed based on the pipeline parameter values:
+** **Lines 22-28**: The `service-1` workflow triggers the `maven/test` job on the `service-1` directory, when the pipeline parameter value mapped to run-build-service-1-job is set to `true`. This will only happen if a change was made in the `serivce-1` directory in the commit, as determined based on the path filtering in the setup phase (`config.yml`).
+** **Lines 31-37**: The `service-2` workflow triggers the `maven/test` job on the `service-2` directory, when the pipeline parameter value mapped to run-build-service-2-job is set to `true`. This will only happen if a change was made in the `serivce-2` directory in the commit, as determined based on the path filtering in the setup phase (`config.yml`).
+** **Lines 43-50**: The `run-integration-tests` workflow will run if the `run-build-service-1-job` or `run-build-service-2-job` pipeline parameters have been updated to `true` based on the results of the path filtering. This runs the `maven/test` job on the `tests` directory to run integration tests against the built services.
+
+#TO DO#
+
+* #Add similar example but showing config splitting. Split the continuation config for the different directories up and store in the relevant dir and have them brought together using the path filtering orb#
+
+* #Add further examples, such as piecing together config elements? https://github.com/ryanpedersen42/config-pack-dynamic/blob/main/.circleci/config.yml#
```

Comment:
Here's an example for sharing config with the path-filtering orb:

## config.yml

Provide path filter mappings with one mapping always being true for the shared config.

```yml
version: 2.1

# define this file as the setup phase of your dynamic configuration
setup: true

# define the parameters that will be used to generate the dynamic configuration.
parameters:
  always-continue:
    type: boolean
    default: false
  build-code:
    type: boolean
    default: false
  build-docs:
    type: boolean
    default: false

# invoke the path-filtering orb to make the filter job available.
orbs:
  path-filtering: circleci/path-filtering@1.0.0

workflows:
  setup-workflow:
    jobs:
      - path-filtering/filter:
          base-revision: main
          mapping: | # The mapping will be used to generate the dynamic configuration for all conditions that match.
            .* always-continue true .circleci/shared-config.yml
            src/.* build-code true .circleci/code-config.yml
            doc/.* build-docs true .circleci/docs-config.yml
```

## shared-config.yml

Define any shared workflows, jobs, parameters, etc.

```yml
version: 2.1

# define the parameters from the setup config.
parameters:
  always-continue:
    type: boolean
    default: false
  build-code:
    type: boolean
    default: false
  build-docs:
    type: boolean
    default: false

# define the shared jobs that will be available for all continued workflows.
jobs:
  lint:
    docker:
      - image: cimg/base:stable
    steps:
      - run: echo "Running linting"

  test:
    docker:
      - image: cimg/base:stable
    steps:
      - run: echo "Running tests"
```

## code-config.yml

Use the shared definitions in the continued workflow.

```yml
version: 2.1

jobs:
  build-code:
    docker:
      - image: cimg/base:stable
    steps:
      - run: echo "Building code"

workflows:
  code-workflow:
    jobs:
      - lint # use the shared lint job.
      - test # use the shared test job.
      - build-code
```

## docs-config.yml

```yml
version: 2.1

jobs:
  build-docs:
    docker:
      - image: cimg/base:stable
    steps:
      - run: echo "Building docs"

workflows:
  docs-workflow:
    jobs:
      - lint # use the shared lint job.
      - build-docs
```

## Notes

- All matched configs are being merged in the path-filtering orb, any name collisions will result in one overriding the other.

Reply:
Thank you!! @liamclarkedev DO we need to include the params in the code and docs config there?

```
parameters:
  always-continue:
    type: boolean
    default: false
  build-code:
    type: boolean
    default: false
  build-docs:
    type: boolean
    default: false
```

Reply:
Also do you think It is also worth me mentioning that you can use partial configs as in [this project](https://github.com/ryanpedersen42/config-pack-dynamic/blob/main/.circleci/config.yml) IIUC and piece them together (pack) and verify with the CLI? I realise the project I linked to is old

Reply:
The parameters can go in either, the shared config is being merged with the code and docs configs so it is the same as defining them in both.

Yes, that sounds like it would be useful, here's the same example as above but with the shared config packed up:

```yml
version: 2.1

# define this file as the setup phase of your dynamic configuration
setup: true

# define the parameters that will be used to generate the dynamic configuration.
parameters:
  always-continue:
    type: boolean
    default: false
  build-code:
    type: boolean
    default: false
  build-docs:
    type: boolean
    default: false

# invoke the orbs to filter, pack and continue configs.
orbs:
  path-filtering: circleci/path-filtering@1.0.0
  circleci-cli: circleci/circleci-cli@0.1.9
  continuation: circleci/continuation@1.0.0

jobs:
  setup:
    executor: path-filtering/default
    steps:
      - checkout
      - circleci-cli/install

      # Generate the shared configuration from a directory with the pack command.
      - run:
          name: Generate shared configuration
          command: circleci config pack .circleci/shared >> .circleci/shared-config.yml

      # The mapping will be used to generate the dynamic configuration for all conditions that match.
      - path-filtering/set-parameters:
          base-revision: main
          mapping: |
            .* always-continue true .circleci/shared-config.yml
            src/.* build-code true .circleci/code-config.yml
            doc/.* build-docs true .circleci/docs-config.yml

      # Generate the dynamic configuration based on the parameters set in the previous step.
      - path-filtering/generate-config

      # Optionally validate the generated configuration.
      - run:
          name: Validate config
          command: circleci config validate /tmp/generated-config.yml

      # Continue the pipeline with the generated configuration.
      - continuation/continue:
          configuration_path: /tmp/generated-config.yml

workflows:
  setup-workflow:
    jobs:
      - setup
```